### ISSUE TITLE
Memory Increase and Runtime Error with react-resizable-panels and GitHub Card Input

### ISSUE THREAD

**Issue by DevGlyph** [toeverything/blocksuite#6471]
Hello BlockSuite Team! 👋

Firstly, I must commend the exceptional work done on the BlockSuite project and the exemplary examples provided by @doodlewind. The clarity and comprehensiveness of these examples have significantly eased the learning curve, making it a joy to explore and integrate BlockSuite into my React project. I also want to express my appreciation for the prompt and helpful responses to issues raised in the past; it's fantastic to see such active and supportive community engagement!

I'm reaching out to report a couple of issues I've encountered while integrating BlockSuite with `react-resizable-panels`. It seems that when wrapping BlockSuite within resizable panels, the memory usage gradually increases during resizing, eventually causing the application to freeze. This issue might stem from resizing a panel, which triggers a React update, performing something quite expensive during rendering. Here's a simplified snippet of how BlockSuite is being used within my application:

```jsx
<EditorProvider>
  <div className="app">
    <PanelGroup direction="horizontal">
      <Panel defaultSize={25}>
        <Sidebar />
      </Panel>
      <PanelResizeHandle
        style={{
          borderLeft: "1px solid #ccc",
        }}
      />
      <Panel defaultSize={75}>
        <div className="main-content">
          <TopBar />
          <EditorContainer />
        </div>
      </Panel>
    </PanelGroup>
  </div>
</EditorProvider>
```
I've created a reproducible CodeSandbox example here: [link](https://codesandbox.io/p/github/DevGlyph/resiable-panels-problem/master?import=true)

Additionally, I've observed a runtime error associated with the GitHub card information input panel. Specifically, when clicking on the input box and then clicking outside to blur the input (repeating this action multiple times), it results in a runtime error. It appears the issue may lie within blocksuite-inline. Here's an image illustrating the error: 

![image](https://github.com/toeverything/blocksuite/assets/162789479/b775d838-e378-46f3-9f41-4db63aef6e89)

I would greatly appreciate any insights or guidance on how to address these issues. Your support has been instrumental in my project's development, and I'm looking forward to finding a resolution with your expertise.

Thank you once again for your incredible work on BlockSuite and for fostering such a helpful community around it. 😀

**Comment by fundon** [comment]
This creates a loop that makes it impossible to get the full path to the parent.

<img width="408" alt="Screenshot 2024-03-13 at 16 39 17" src="https://github.com/toeverything/blocksuite/assets/27926/3cd503cb-c4d0-44f7-8e37-a986c090ba47">

https://github.com/bvaughn/react-resizable-panels/blob/44d3a65d649dd943168e581fe400c774cfc06cfc/packages/react-resizable-panels/src/vendor/stacking-order.ts#L132

https://github.com/bvaughn/react-resizable-panels/blob/44d3a65d649dd943168e581fe400c774cfc06cfc/packages/react-resizable-panels/src/vendor/stacking-order.ts#L117-L133

**Comment by bvaughn** [comment]
The change shown above can't be made without breaking other functionality. Can you please confirm if this change also fixes the issue you're reporting?

```diff
diff --git a/packages/react-resizable-panels/src/vendor/stacking-order.ts b/packages/react-resizable-panels/src/vendor/stacking-order.ts
index f3dd291..347da05 100644
--- a/packages/react-resizable-panels/src/vendor/stacking-order.ts
+++ b/packages/react-resizable-panels/src/vendor/stacking-order.ts
@@ -120,7 +120,14 @@ function get_ancestors(node: HTMLElement) {
 
   while (node) {
     ancestors.push(node);
-    node = get_parent(node);
+
+    const parent = get_parent(node);
+
+    if (node === parent) {
+      break;
+    }
+
+    node = parent;
   }
 
   return ancestors; // [ node, ... <body>, <html>, document ]
```

**Comment by DevGlyph** [comment]
Dear @fundon  and @bvaughn,

Firstly, I would like to extend my deepest gratitude for your swift responses and for looking into the issue I've encountered. Your dedication to the community and the support you provide are truly commendable.

As a newcomer to front-end development, I have been following your discussion closely and took the initiative to delve into some breakpoint debugging using the browser console to better understand the situation. Based on the insights gained from your conversation and my debugging efforts, it seems that the proposed fix might not address the core of the issue.

Regarding the repair code suggested in the react-resizable-panels repository, here’s my observation:

To elaborate, during the execution of `get_parent()` when the `node` is `div.page-editor-container`, the `node.parentNode.host` returns the result of blocksuit’s `PageEditor.host()` which is `editor-host`. However, when the `node` is `editor-host`, the `get_parent()` finds that `node.parentNode.host` does not exist and ends up returning `node.parentNode` which is `div.page-editor-container` again. This seems to create a loop where `node` and `parent` will never be equal, hence the infinite loop persists.

![image](https://github.com/toeverything/blocksuite/assets/162789479/cd8d3fb0-4a31-4a6d-b2dc-a7b1b55ca865)

![image](https://github.com/toeverything/blocksuite/assets/162789479/f7941988-3e48-4378-80af-fa0ff99a9ab9)

![image](https://github.com/toeverything/blocksuite/assets/162789479/9c7b04c8-f6fa-4e6b-be93-ed6d51a00f77)

I’m still learning, and your expertise is incredibly valuable to someone just starting out like myself. I would be grateful for any additional guidance you could provide on this matter.

Lastly, @fundon, if I may ask, what tool did you use to pinpoint the issue within the `stacking-order` library? As a beginner, I am very curious about the professional tools that experienced developers like you use. I would love to learn more and try them out myself. 😶

Thank you all once again for your time and patience. ❤️

**Comment by bvaughn** [comment]
@DevGlyph you're welcome!

Once either you or @fundon have verified the [change I posted above](https://github.com/toeverything/blocksuite/issues/6471#issuecomment-1994364302), I'll publish a release. (and if that change doesn't fix things, we'll try somethign else)

**Comment by fundon** [comment]
@bvaughn Thanks for your support.


> The change shown above can't be made without breaking other functionality. Can you please confirm if this change also fixes the issue you're reporting?

It doesn't work.

I tried the following modification and it works fine, you can review it.

```
diff --git a/packages/react-resizable-panels/src/vendor/stacking-order.ts b/packages/react-resizable-panels/src/vendor/stacking-order.ts
index f3dd291..968931c 100644
--- a/packages/react-resizable-panels/src/vendor/stacking-order.ts
+++ b/packages/react-resizable-panels/src/vendor/stacking-order.ts
@@ -128,6 +128,13 @@ function get_ancestors(node: HTMLElement) {

 /** @param {HTMLElement} node */
 function get_parent(node: HTMLElement) {
-  // @ts-ignore
-  return node.parentNode?.host || node.parentNode;
+  const { parentNode } = node;
+  if (parentNode) {
+    // @ts-ignore
+    const { host } = parentNode;
+    if (host && !parentNode.contains(host)) {
+      return host
+    }
+  }
+  return parentNode;
 }
```

**Comment by fundon** [comment]
@bvaughn Welcome!

> Lastly, @fundon, if I may ask, what tool did you use to pinpoint the issue within the `stacking-order` library? As a beginner, I am very curious about the professional tools that experienced developers like you use. I would love to learn more and try them out myself. 😶

I usually try to debug in Firefox or Chrome. When I read this issue. 

1. Try dragging the separator, 

<img width="780" alt="Screenshot 2024-03-14 at 08 28 58" src="https://github.com/toeverything/blocksuite/assets/27926/3735d328-2622-42b5-9363-bfa537e1eefd">

2. Next.js throws an error
<img width="1425" alt="Screenshot 2024-03-14 at 08 33 27" src="https://github.com/toeverything/blocksuite/assets/27926/e4fca661-9aa7-4a39-b792-992516e965f6">

`RangeError: invalid array length`: what is it? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Invalid_array_length

So I'm guessing there's a problem with the `get_parent` method.

3. Start tracking `get_parent` method and make some changes.

4. As in your screenshot, I also use the debugger tool to watch `parentNode` and `host`

I hope the above helps you.

**Comment by bvaughn** [comment]
> I tried the following modification and it works fine, you can review it.

I don't think that looks sound for general usage. It seems like it's pretty shadow root specific.

Can you trim this bug down to a minimal repro that breaks the stacking-order library (that doesn't involve loading the complete editor)?

**Comment by fundon** [comment]
> @fundon I'm not sure if that's sound for general usage. I'm concerned this workaround feels pretty shadow-root specific.

Yes, there is no real solution to the problem of shadow-root.

> **Edit** That change also breaks the `react-resizable-panels` e2e tests for what it's worth.

I can submit a PR and add some tests.

**Comment by bvaughn** [comment]
Happy to review a PR, sure. So long as the change doesn't break current functionality.

**Comment by bvaughn** [comment]
If you update to react-resizable-panels@2.0.13, this should be fixed

**Comment by DevGlyph** [comment]
Hey everyone,

Just wanted to drop in quickly - I've been tied up with some coursework, not a "school project" (didn't mean to sound like I'm clocking in hours for school)

Huge props for the fix on the react-resizable-panels memory leak. Major thanks to @fundon for not only tackling the issue but also taking the time to walk me through the diagnosis process with detailed explanations and visuals. Getting to see that level of problem-solving up close is super valuable, especially for someone still learning the ropes like me. 

I took the updates for a test drive, and everything's running like clockwork. Also, I couldn’t replicate the GitHub card issue anymore (Error: Permission denied to access property "contains" in blocksuite-inline). Seems like we're in the clear on both fronts!

Props to everyone involved for their hard work and for sharing those insights. It's super helpful for us newbies, and I learned a ton from the breakdown of the fix.

I feel confident that both issues have been resolved, so I'm going to go ahead and close this issue now. I really can't thank you all enough for your help and guidance throughout this process. You've made a significant impact on my learning journey, and I'm extremely grateful for it.

Keep rocking the open-source world! 🎸
